File: android/Detonator/src/main/java/com/iconshot/detonator/module/StorageModule.java
package com.iconshot.detonator.module;

import android.content.Context;
import android.content.SharedPreferences;

import com.iconshot.detonator.Detonator;

public class StorageModule extends Module {
    public StorageModule(Detonator detonator) {
        super(detonator);
    }

    @Override
    public void setUp() {
        detonator.setRequestListener("com.iconshot.detonator.storage::getItem", (promise, value, edge) -> {
            GetItemData data = detonator.decode(value, GetItemData.class);

            SharedPreferences sharedPreferences = getSharedPreferences(data.name);

            String itemValue = sharedPreferences.getString(data.key, null);

            promise.resolve(itemValue != null ? "&" + itemValue : ":n");
        });

        detonator.setRequestListener("com.iconshot.detonator.storage::setItem", (promise, value, edge) -> {
            SetItemData data = detonator.decode(value, SetItemData.class);

            SharedPreferences sharedPreferences = getSharedPreferences(data.name);

            SharedPreferences.Editor editor = sharedPreferences.edit();

            editor.putString(data.key, data.value);

            editor.apply();

            promise.resolve();
        });
    }

    private SharedPreferences getSharedPreferences(String name) {
        return detonator.context.getSharedPreferences(name, Context.MODE_PRIVATE);
    }

    private static class GetItemData {
        String name;
        String key;
    }

    private static class SetItemData {
        String name;
        String key;
        String value;
    }
}

File: ios/Detonator/Sources/Detonator/Module/StorageModule/StorageModule.swift
class StorageModule: Module {
    private var storages: [String: Storage] = [:]
    
    private func getStorage(name: String) -> Storage {
        if let storage = storages[name] {
            return storage
        }
        
        let storage = Storage(name: name)
        
        storages[name] = storage
        
        return storage
    }
    
    override func setUp() -> Void {
        detonator.setRequestListener("com.iconshot.detonator.storage::getItem") { promise, value, edge in
            let data: GetItemData = self.detonator.decode(value)!
            
            let storage = self.getStorage(name: data.name)
            
            let itemValue = storage.content[data.key]
            
            promise.resolve(itemValue != nil ? "&\(itemValue)" : ":n")
        }
        
        detonator.setRequestListener("com.iconshot.detonator.storage::setItem") { promise, value, edge in
            let data: SetItemData = self.detonator.decode(value)!
            
            let storage = self.getStorage(name: data.name)
            
            storage.content[data.key] = data.value
            
            promise.resolve()
        }
    }
    
    private struct GetItemData: Decodable {
        let name: String
        let key: String
    }
    
    private struct SetItemData: Decodable {
        let name: String
        let key: String
        let value: String
    }
}
